### PR TITLE
validate ipv4 ipv6 algorithms in C

### PR DESCRIPTION
**Pull request by kobra-cai**
**Fixes issue:**
#3431 
Added validate IPv6 and IPv4 (2 separate but really similar files) in C

**Changes:**
Implemented two separate but similar files called validate_ipv4.c and validate_ipv6 that validate both types respectively. They both, however, have very similar code with very small minor differences and I can try to do one of them in a completely different way if needed.

### CODE REVIEW

**Comment by kobra-cai** [comment]
Fixed my original pull request issue (info can be found here: https://github.com/OpenGenus/cosmos/pull/3494)

File: code/networking/src/validate_IP/validate_ipv4.c [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+char 
+*is_ipv4(char *ip_addr);
+
+char 
+*is_ipv4(char *ip_addr) 
+{
+	char *ans = "";
+	struct sockaddr_in sa;
+	
+	if(inet_pton(AF_INET, ip_addr, &(sa.sin_addr))) {
```
Comment (arnavb) [inline code/networking/src/validate_IP/validate_ipv4.c]: Ternary?

File: code/networking/src/validate_IP/validate_ipv4.c [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <arpa/inet.h>
+
+char 
+*is_ipv4(char *ip_addr);
+
+char
```
Comment (arnavb) [inline code/networking/src/validate_IP/validate_ipv4.c]: Our style guide says its `char*`, not `*is_ipv4`.

File: code/networking/src/validate_IP/validate_ipv4.c [hunk 3]
```diff
@@ -0,0 +1,62 @@
+/* This implementation reads in a file as argv[1] and checks if each line of the file is a valid IPv4 address.
+Algorithm of the IPv4 validation is in the char *is_ipv4(char *ip_addr) function.
+
+Implementation utilized inet_pton() and information about that function 
+from https://beej.us/guide/bgnet/html/multi/inet_ntopman.html */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <arpa/inet.h>
```
Comment (arnavb) [inline code/networking/src/validate_IP/validate_ipv4.c]: This file _is_ installed, right?

File: code/networking/src/validate_IP/validate_ipv4.c [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+		} else {
+			ip_addr = (char *) malloc(n);
+			strncpy(ip_addr, line, n);
+		}
+		char* ans = is_ipv4(ip_addr);
+		printf("Is '%s' valid IPv4? \n%s\n", ip_addr, ans);
+	}
+
+	fclose(fp);
+	return 0;
```
Comment (arnavb) [inline code/networking/src/validate_IP/validate_ipv4.c]: Should be in parentheses.

File: code/networking/src/validate_IP/validate_ipv4.c [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+}
```
Comment (arnavb) [inline code/networking/src/validate_IP/validate_ipv4.c]: Add a newline, please.

File: code/networking/src/validate_IP/validate_ipv4.c [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+	while ((n = getline(&line, &size, fp)) != -1) {
+		if(line[n-1] == '\n') {
+			line[n-1] = '\0';
+			ip_addr = (char *) malloc(n-2);
+			strncpy(ip_addr, line, n-2);
+		} else {
+			ip_addr = (char *) malloc(n);
+			strncpy(ip_addr, line, n);
+		}
+		char* ans = is_ipv4(ip_addr);
```
Comment (arnavb) [inline code/networking/src/validate_IP/validate_ipv4.c]: No need for the variable.

File: code/networking/src/validate_IP/validate_ipv4.c [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+	const char *filename = argv[1];
+	fp = fopen(filename, "r");
+
+	if (!fp) {
+		fprintf(stderr, "Error in opening the file!\n");
+		exit(1);
+	}
+
+	while ((n = getline(&line, &size, fp)) != -1) {
+		if(line[n-1] == '\n') {
```
Comment (arnavb) [inline code/networking/src/validate_IP/validate_ipv4.c]: Space before `(`.

File: code/networking/src/validate_IP/validate_ipv4.c (hunk 3 above)
Comment (kobra-cai) [reply code/networking/src/validate_IP/validate_ipv4.c]: Just a header you can include, at least in UNIX systems. It seems that in Windows, you have to use Winsock2.h

**Comment by kobra-cai** [comment]
Made the suggested changes in my last commit for this pull request

**Comment by kobra-cai** [comment]
Realized I still forgot to add the newline at the end of the file, I'll get fix that tonight or tomorrow!

**Comment by AdiChat** [comment]
Looks great 😍 

Hope you are enjoying your journey 🚆  with **OpenGenus**
positive_opengenus_first

**Comment by kobra-cai** [comment]
Thank you!

**Pull request merged**